util/encryption.go

```diff
@@ -0,0 +1,132 @@
+package util
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+)
+
+var (
+	// "the key" to encrypt/decrypt passwords with
+	key        = os.Getenv("ENCRYPTION_KEY")
+	keyPresent = false
+
+	// regex to pull the actual secret out of the miq-password style column
+	xtractRe = regexp.MustCompile(`v2:{(.+)}`)
+)
+
+// init for this file basically just decodes the ENCRYPTION_KEY env var to the
+// plain text equivalent
+func init() {
+	if key == "" {
+		return
+	}
+
+	// the key is base64 encoded in the ENV
+	decoded, err := base64.RawStdEncoding.DecodeString(string(key))
+	if err != nil {
+		panic(err)
+	}
+	key = string(decoded)
+	keyPresent = true
+}
+
+// Encrypts str into a miq-compatible password format using the encryption key
+// in the environment
+func Encrypt(str string) (string, error) {
+	if !keyPresent {
+		return "", fmt.Errorf("no encryption key present")
+	}
+
+	encoded, err := encode(str)
+	if err != nil {
+		return "", err
+	}
+
+	// base64 encode the encrypted secret for text-storage
+	return "v2:{" + base64.RawStdEncoding.EncodeToString([]byte(encoded)) + "}", nil
+}
+
+// Decrypts a miq-compatible password into a string
+func Decrypt(str string) (string, error) {
+	if !keyPresent {
+		return "", fmt.Errorf("no encryption key present")
+	}
+
+	if !xtractRe.Match([]byte(str)) {
+		return "", fmt.Errorf("string does not match format")
+	}
+
+	// extract the password from between the brackets
+	matches := xtractRe.FindAllStringSubmatch(str, 1)
+
+	// the password is base64 encoded
+	rawPass, err := base64.RawStdEncoding.DecodeString(matches[0][1])
+	if err != nil {
+		return "", err
+	}
+
+	return decode(string(rawPass))
+}
+
+func decode(pw string) (string, error) {
+	// create the block from the key
+	block, err := aes.NewCipher([]byte(key))
+	if err != nil {
+		return "", err
+	}
+
+	// no iv used, but need an iv that is the same length as the block size.
```

Comment:
Hi 😁  please what does "iv" mean?

Reply:
Without going into _too_ much details the IV is kind of like a salt on the string for encryption.

The way it **differs** from a salt is this:
- IV keys should be unique per record (e.g. randomly generated, then stored alongside for decryption)
- they _aren't_ secret, it is safe to store them in plaintext right alongside the cipher.

To decrypt with an IV you just provide it sort of as a second key. In the miq-password gem it looks like they don't use IV's which is kind of a no-no, we're not using that extra layer of security. I'm sort-of glad we can move away from it eventually through vault or we could eventually change out our implementation to use an IV per-password.

More in-depth explanation here: https://stackoverflow.com/a/39412887